contracts/shapella-upgrade/ShapellaUpgradeTemplate.sol

```diff
@@ -0,0 +1,546 @@
+
+        _initializeStakingRouter();
+
+        _migrateDSMGuardians();
+    }
+
+    function _assertCorrectInitialState() internal view {
+        if (ILidoOracle(address(_legacyOracle())).getVersion() != 3) revert LidoOracleMustNotBeUpgradedToLegacyYet();
+
+        _assertAdminsOfProxies(address(this));
+
+        _assertInitialDummyProxyImplementations();
+
+        // Check roles of non-proxy contracts (can do without binding implementations)
+        _assertSingleOZRoleHolder(_hashConsensusForAccountingOracle, DEFAULT_ADMIN_ROLE, address(this));
+        _assertSingleOZRoleHolder(_hashConsensusForValidatorsExitBusOracle, DEFAULT_ADMIN_ROLE, address(this));
+        _assertSingleOZRoleHolder(_burner(), DEFAULT_ADMIN_ROLE, address(this));
+        if (_depositSecurityModule().getOwner() != address(this)) revert WrongDsmOwner();
+    }
+
+    function _upgradeProxyImplementations() internal {
+        _accountingOracle().proxy__upgradeTo(_accountingOracleImplementation);
+        _validatorsExitBusOracle().proxy__upgradeTo(_validatorsExitBusOracleImplementation);
+        _stakingRouter().proxy__upgradeTo(_stakingRouterImplementation);
+        _withdrawalVault().proxy__upgradeTo(_withdrawalVaultImplementation);
+        _withdrawalQueue().proxy__upgradeTo(_withdrawalQueueImplementation);
+    }
+
+    function _assertAdminsOfProxies(address admin) internal view {
+        _assertProxyAdmin(_accountingOracle(), admin);
+        _assertProxyAdmin(_locator, admin);
+        _assertProxyAdmin(_stakingRouter(), admin);
+        _assertProxyAdmin(_validatorsExitBusOracle(), admin);
+        _assertProxyAdmin(_withdrawalQueue(), admin);
+        _assertProxyAdmin(_withdrawalVault(), admin);
+    }
+
+    function _assertProxyAdmin(IOssifiableProxy proxy, address admin) internal view {
+        if (proxy.proxy__getAdmin() != admin) revert WrongProxyAdmin(address(proxy));
+    }
+
+    function _assertInitialDummyProxyImplementations() internal view {
+        _assertInitialDummyImplementation(_accountingOracle());
+        _assertInitialDummyImplementation(_stakingRouter());
+        _assertInitialDummyImplementation(_validatorsExitBusOracle());
+        _assertInitialDummyImplementation(_withdrawalQueue());
+        _assertInitialDummyImplementation(_withdrawalVault());
+    }
+
+    function _assertInitialDummyImplementation(IOssifiableProxy proxy) internal view {
+        if (proxy.proxy__getImplementation() != _dummyImplementation) revert WrongInitialImplementation(address(proxy));
+    }
+
+    function _assertSingleOZRoleHolder(IAccessControlEnumerable accessControlled, bytes32 role, address holder) internal view {
+        if (accessControlled.getRoleMemberCount(role) != 1
+         || accessControlled.getRoleMember(role, 0) != holder
+        ) {
+            revert WrongSingleRoleHolder(address(accessControlled), role);
+        }
+    }
+
+    function _assertCorrectNonAdminRoleHolders() internal view {
```

Comment:
Shall we also check/grant other roles of the contracts? Maybe offchain if gas or contract limit is reached.

Reply:
done